src/Python/Search.cc

```diff
@@ -0,0 +1,106 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.hltpr.rwth-aachen.de/rwth-asr/rwth-asr-license.html
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "Search.hh"
+
+#include <Search/Module.hh>
+#include <Speech/ModelCombination.hh>
+
+namespace py = pybind11;
+
+SearchAlgorithm::SearchAlgorithm(const Core::Configuration& c)
+        : Core::Component(c),
+          searchAlgorithm_(Search::Module::instance().createSearchAlgorithmV2(select("search-algorithm"))) {
+    searchAlgorithm_->setModelCombination({config, searchAlgorithm_->requiredModelCombination(), searchAlgorithm_->requiredAcousticModel()});
+}
+
+void SearchAlgorithm::reset() {
+    searchAlgorithm_->reset();
+}
+
+void SearchAlgorithm::enterSegment() {
+    searchAlgorithm_->enterSegment();
+}
+
+void SearchAlgorithm::finishSegment() {
+    searchAlgorithm_->finishSegment();
+}
+
+void SearchAlgorithm::putFeature(py::array_t<f32> const& feature) {
+    size_t F = 0ul;
+    if (feature.ndim() == 2) {
+        if (feature.shape(0) != 1) {
+            error() << "Received feature tensor with non-trivial batch dimension " << feature.shape(0) << "; should be 1";
+        }
+        F = feature.shape(1);
+    }
+    else if (feature.ndim() != 1) {
+        error() << "Received feature vector of invalid dim " << feature.ndim() << "; should be 1";
+        F = feature.shape(0);
+    }
+
+    searchAlgorithm_->putFeature({feature, F});
+}
+
+void SearchAlgorithm::putFeatures(py::array_t<f32> const& features) {
+    size_t T = 0ul;
+    size_t F = 0ul;
+    if (features.ndim() == 3) {
+        if (features.shape(0) != 1) {
+            error() << "Received feature tensor with non-trivial batch dimension " << features.shape(0) << "; should be 1";
+        }
+        T = features.shape(1);
+        F = features.shape(2);
+    }
+    else if (features.ndim() != 2) {
+        error() << "Received feature tensor of invalid dim " << features.ndim() << "; should be 2 or 3";
+        T = features.shape(0);
+        F = features.shape(1);
+    }
+
+    for (size_t t = 0ul; t < T; ++t) {
+        searchAlgorithm_->putFeatures({features, T * F}, T);
+    }
+}
+
+Traceback SearchAlgorithm::getCurrentBestTraceback() {
+    searchAlgorithm_->decodeManySteps();
+
+    auto                       traceback = searchAlgorithm_->getCurrentBestTraceback();
+    std::vector<TracebackItem> result;
+    result.reserve(traceback->size());
+
+    u32 prevTime = 0;
+
+    for (auto it = traceback->begin(); it != traceback->end(); ++it) {
+        result.push_back({
+                it->pronunciation->lemma()->symbol(),
```

Comment:
```suggestion
    for (auto it = traceback->begin(); it != traceback->end(); ++it) {
        std::string lemma = "";
        if (it->pronunciation and it->pronunciation->lemma()) {
            lemma = std::string(it->pronunciation->lemma()->symbol().str());
        }
        result.push_back({
                lemma,
```

This caused a segmentation fault